kotlin-result-coroutines/src/commonMain/kotlin/com/github/michaelbull/result/coroutines/binding/SuspendableBinding.kt

```diff
@@ -3,60 +3,78 @@ package com.github.michaelbull.result.coroutines.binding
 import com.github.michaelbull.result.Err
 import com.github.michaelbull.result.Ok
 import com.github.michaelbull.result.Result
-import kotlinx.coroutines.CancellationException
-import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.cancel
+import kotlinx.coroutines.async
 import kotlinx.coroutines.coroutineScope
 import kotlinx.coroutines.sync.Mutex
 import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.cancel
 import kotlin.contracts.InvocationKind
 import kotlin.contracts.contract
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.resume
 
 /**
  * Suspending variant of [binding][com.github.michaelbull.result.binding].
- * Wraps the suspendable block in a new coroutine scope.
- * This scope is cancelled once a failing bind is encountered, allowing deferred child jobs to be eagerly cancelled.
+ * Wraps the suspendable block in an async job, inheriting the parent coroutine context.
+ * This async job is cancelled once a failing bind is encountered, eagerly cancelling all children.
  */
 public suspend inline fun <V, E> binding(crossinline block: suspend SuspendableResultBinding<E>.() -> V): Result<V, E> {
     contract {
         callsInPlace(block, InvocationKind.EXACTLY_ONCE)
     }
-    val receiver = SuspendableResultBindingImpl<E>()
-
-    return try {
-        coroutineScope {
-            receiver.coroutineScope = this@coroutineScope
+    lateinit var receiver: SuspendableResultBindingImpl<E>
+    return coroutineScope {
+        val parentJob = async {
```

Comment:
Why do we need to call async when we're already in a coroutine context? This will end up creating a new coroutine which I'm not sure is intentional. I think we should be using `withContext` here instead.

Reply:
believe me I tried every permutation under the sun before coming to this conclusion, even ended up on the coroutine slack to verify the behaviour I was trying to achieve. Check this gist I ended up creating that someone kindly helped out with: https://gist.github.com/Munzey/212f38879039987a4c819c53d1d393cd

Basically without wrapping the block in an `async` coroutine, you can't catch the `BindCancellationException`